### PR TITLE
Implement validation of best practices for social networks sharing

### LINKED ISSUE #128: Implement content validation
In order to increase content quality we need to implement validation of the content to be shared. The validator needs to take into account best practices as dictated mainly by [Twitter](https://dev.twitter.com/cards/markup) and [Facebook](https://developers.facebook.com/docs/sharing/best-practices):

* content title should not have more than 70 characters
* content description should not have more than 200 characters and should contain at least 2 phrases
* lead image should be in JPG, PNG, WEBP or GIF format
* lead image dimensions should be at least 600px x 315px and aspect ratio should be 1.33:1 at least
* lead image size should be less than 5MB

Validation must be executed on content publishing or when published content is edited. Validation is not supposed to enforce these practices, but only warn editors (using a status message) about failure to comply with them.

### PR DESCRIPTION
**Pull request by claytonc**
closes #128

### CODE REVIEW

File: sc/social/like/configure.zcml [hunk 1]
```diff
@@ -34,6 +34,34 @@
...
+      />
+
+  <subscriber
+      zcml:condition="installed plone.app.dexterity"
+      for="plone.dexterity.interfaces.IDexterityContent
+           plone.dexterity.interfaces.IEditFinishedEvent"
+      handler=".subscribers.social_content_check"
+      />
+
+  <subscriber
```
Comment (hvelarde) [inline sc/social/like/configure.zcml]: to simplify development this new feature can be implemented for Dexterity-based content types only.

File: sc/social/like/configure.zcml [hunk 2]
```diff
@@ -34,6 +34,34 @@
...
+      zcml:condition="installed Products.ATContentTypes"
+      for="Products.ATContentTypes.interfaces.IATContentType
+           Products.DCWorkflow.interfaces.IAfterTransitionEvent"
+      handler=".subscribers.social_content_check_workflow"
+      />
+
+  <subscriber
```
Comment (hvelarde) [inline sc/social/like/configure.zcml]: same here: Dexterity-based content types only.

File: sc/social/like/subscribers.py [hunk 3]
```diff
@@ -107,3 +113,49 @@ def assign_canonical_url(obj, event):
             'Canonical domain not set in Social Media configlet; '
             "Facebook's Open Graph canonical URL (og:orl) will not be available"
         )
+
+
+def social_content_check(obj, event):
+    """Check if content is valid if enabled in sc.social.content, in action when edit."""
+
+    request = obj.REQUEST
+    try:
```
Comment (hvelarde) [inline sc/social/like/subscribers.py]: can we use the add-on browser layer instead?

File: sc/social/like/utils.py [hunk 4]
```diff
@@ -112,3 +112,65 @@ def get_valid_objects(brains):
...
+        msg = u'Image aspect ratio should be 1.33:1 at least.'
+        logger.info(msg)
+        return msg
+
+
+def get_ratio(w, h):
+    """Calculate aspect ratio."""
+    w = float(w)
+    h = float(h)
+    r = w % h or w
```
Comment (hvelarde) [inline sc/social/like/utils.py]: I didn't understood why you need this.

File: sc/social/like/utils.py [hunk 5]
```diff
@@ -112,3 +112,65 @@ def get_valid_objects(brains):
...
+        return msg.format(type)
+
+    size = value.size
+    if size > 5242880:
+        msg = u'Image size should be less than 5MB.'
+        logger.info(msg)
+        return msg
+
+    width, height = (value.width, value.height)
+    if width < 600 and height < 315:
```
Comment (hvelarde) [inline sc/social/like/utils.py]: this seems to be wrong; you should use `or` instead.

File: sc/social/like/utils.py [hunk 6]
```diff
@@ -112,3 +112,65 @@ def get_valid_objects(brains):
...
+
+    return
+
+
+def validate_image_social(value):
+    """Check if image be in formats mime type, dimensions and size."""
+
+    list_mimetypes = ['image/jpeg', 'image/png', 'image/gif', ' image/webp']
+
+    type = getattr(value, 'mimetype', None)
```
Comment (hvelarde) [inline sc/social/like/utils.py]: this should work for Dexterity-based content types only

File: sc/social/like/configure.zcml (hunk 1 above)
Comment (claytonc) [reply sc/social/like/configure.zcml]: ok

File: sc/social/like/configure.zcml (hunk 2 above)
Comment (claytonc) [reply sc/social/like/configure.zcml]: ok

File: sc/social/like/utils.py (hunk 6 above)
Comment (claytonc) [reply sc/social/like/utils.py]: ok

File: sc/social/like/utils.py (hunk 5 above)
Comment (claytonc) [reply sc/social/like/utils.py]: ok

File: sc/social/like/subscribers.py (hunk 3 above)
Comment (claytonc) [reply sc/social/like/subscribers.py]: I have not tried I used base the 'amp'

File: sc/social/like/subscribers.py [hunk 7]
```diff
@@ -107,3 +114,50 @@ def assign_canonical_url(obj, event):
...
+
+def social_content_check(obj, event):
+    """Check if content is valid if enabled in sc.social.content, in action when edit."""
+
+    request = getRequest()
+
+    try:
+        view = getMultiAdapter((obj, request), name='social_likes_view')
+    except ComponentLookupError:
+        logger.info('not instaled.')
```
Comment (idgserpro) [inline sc/social/like/subscribers.py]: I would change the message to 'social_likes_view not installed'. Having a string 'not installed' without other information isn't too much useful...

File: sc/social/like/subscribers.py [hunk 8]
```diff
@@ -107,3 +114,50 @@ def assign_canonical_url(obj, event):
...
+                api.portal.show_message(message=msg, request=request, type='warning')
+        except AttributeError:
+            pass
+
+        image = get_content_image(obj)
+        if image:
+            v_image = validate_image_social(image)
+            if v_image:
+                msg = v_image
+                logger.info(msg)
```
Comment (idgserpro) [inline sc/social/like/subscribers.py]: Why do you use `logger.info` here but no in the `title` and `description` checks above?

File: sc/social/like/utils.py [hunk 9]
```diff
@@ -112,3 +112,67 @@ def get_valid_objects(brains):
             logger.warn(msg.format(b.getPath()))
             continue
         yield obj
+
+
+def validate_title_social(value):
+    """Check if title field have more than 70 characters."""
+    if len(value) > 70:
+        msg = u'Title have more than 70 characters.'
```
Comment (idgserpro) [inline sc/social/like/utils.py]: We need i18n here and in all `msg` variables presented in `utils.py`.

File: sc/social/like/subscribers.py (hunk 8 above)
Comment (hvelarde) [reply sc/social/like/subscribers.py]: I prefer `logger.debug()`

File: sc/social/like/subscribers.py (hunk 7 above)
Comment (hvelarde) [reply sc/social/like/subscribers.py]: that's not needed at all; just do nothing and return.

File: sc/social/like/tests/test_subscribers.py [hunk 10]
```diff
@@ -10,13 +9,18 @@
...
 from sc.social.like.interfaces import ISocialLikeSettings
+from sc.social.like.testing import HAS_DEXTERITY
 from sc.social.like.testing import INTEGRATION_TESTING
 from sc.social.like.testing import load_image
 from sc.social.like.tests.api_hacks import set_image_field
+from sc.social.like.tests.utils import get_random_string
+from sc.social.like.utils import HELP_OG_DESCRIPTION_VALIDATOR
+from sc.social.like.utils import HELP_OG_IMAGE_ASPECT_RATIO_VALIDATOR
+from sc.social.like.utils import HELP_OG_IMAGE_DIMENSIONS_VALIDATOR
+from sc.social.like.utils import HELP_OG_TITLE_VALIDATOR
```
Comment (rodfersou) [inline sc/social/like/tests/test_subscribers.py]: isn't it possible to move these constants into config.py file?

File: sc/social/like/subscribers.py [hunk 11]
```diff
@@ -107,3 +113,40 @@ def assign_canonical_url(obj, event):
             'Canonical domain not set in Social Media configlet; '
             "Facebook's Open Graph canonical URL (og:orl) will not be available"
         )
+
+
+def social_content_check(obj, event):
+    """Check if content is valid if enabled in sc.social.content, in action when edit."""
+    request = getRequest()
+
+    # FIXME: view still available with package uninstalled; why?
```
Comment (idgserpro) [inline sc/social/like/subscribers.py]: Is this doc really necessary? Does this happen only in this view or in another Plone views? Do this behavior harms the implementation in any way? If not, this comment can be removed and a new issue opened at https://github.com/plone/Products.CMFPlone/issues for generic views.

File: sc/social/like/tests/utils.py [hunk 12]
```diff
@@ -13,3 +13,9 @@ def enable_social_media_behavior():
     fti.behaviors = tuple(behaviors)
     # invalidate schema cache
     notify(SchemaInvalidatedEvent('News Item'))
+
+
+def get_random_string(lenght):
```
Comment (idgserpro) [inline sc/social/like/tests/utils.py]: Typo, use length for the variable name.

File: sc/social/like/subscribers.py (hunk 11 above)
Comment (claytonc) [inline sc/social/like/subscribers.py]: The error occurs when you install a Plone site.
I also tested `zcml:condition=" installed sc.social.like`, but it did not work.
I also tested `zcml:condition="have sc.social.like`, but it did not work.

Comment (claytonc) [inline sc/social/like/subscribers.py]: Removing this code will cause the `ComponentLookupError` error when installing a Plone site with  plone.app.contenttypes

File: sc/social/like/upgrades/v3047/configure.zcml [hunk 13]
```diff
@@ -0,0 +1,18 @@
+<configure
+    xmlns="http://namespaces.zope.org/zope"
+    xmlns:genericsetup="http://namespaces.zope.org/genericsetup">
+
+  <genericsetup:upgradeSteps
```
Comment (idgserpro) [inline sc/social/like/upgrades/v3047/configure.zcml]: Are you re-running the default profile here?

Comment (hvelarde) [reply sc/social/like/upgrades/v3047/configure.zcml]: no, this is related with https://github.com/zopefoundation/Products.GenericSetup/issues/7

**Pull request merged**